jax/experimental/optix.py

```diff
@@ -0,0 +1,361 @@
+    updates = tree_multimap(
+        lambda m, v: m / (jnp.sqrt(v) + eps), mu_hat, nu_hat)
+    return updates, ScaleByAdamState(count=state.count + 1, mu=mu, nu=nu)
+
+  return init_fn, update_fn
+
+
+ScaleState = collections.namedtuple("ScaleState", "")
+
+
+def scale(step_size):
+  """Scale updates by some fixed scalar `step_size`.
+
+  Args:
+    step_size: a scalar corresponding to a fixed scaling factor for updates.
+
+  Returns:
+    An (init_fn, update_fn) tuple.
+  """
+
+  def init_fn(_):
+    return ScaleState()
+
+  def update_fn(updates, state):
+    updates = tree_multimap(lambda g: step_size * g, updates)
+    return updates, state
+
+  return init_fn, update_fn
+
+
+ScaleByScheduleState = collections.namedtuple("ScaleByScheduleState", "count")
+
+
+def scale_by_schedule(step_size_fn):
+  """Scale updates using a custom schedule for the `step_size`.
+
+  Args:
+    step_size_fn: a function that takes an update count as input and proposes
+      the step_size to multiply the updates by.
+
+  Returns:
+    An (init_fn, update_fn) tuple.
+  """
+
+  def init_fn(_):
+    return ScaleByScheduleState(count=jnp.zeros([]))
+
+  def update_fn(updates, state):
+    updates = tree_multimap(lambda g: step_size_fn(state.count) * g, updates)
+    return updates, ScaleByScheduleState(count=state.count + 1)
+
+  return init_fn, update_fn
+
+
+### Utilities for building and using custom optimizers. ###
+
+
+def chainer(*args):
+  """Applies a list of chainable update transformations.
+
+  Given a sequence of chainable transforms, `chainer` returns an `init_fn`
+  that constructs a `state` by concatenating the states of the individual
+  transforms, and returns an `update_fn` which chains the update transformations
+  feeding the appropriate state to each.
+
+  Args:
+    *args: a sequence of chainable (init_fn, update_fn) tuples.
+
+  Returns:
+    A single (init_fn, update_fn) tuple.
+  """
+
+  init_fns, update_fns = zip(*args)
+
+  def init(params):
+    return [fn(params) for fn in init_fns]
+
+  def update(updates, state):
+    new_state = []
+    for s, fn in zip(state, update_fns):
+      updates, new_s = fn(updates, s)
+      new_state.append(new_s)
+    return updates, new_state
+
+  return init, update
+
+
+def apply_updates(params, updates):
+  """Applies an update to the corresponding parameters.
+
+  This is an (optional) utility functions that applies an update, and returns
+  the updated parameters to the caller. The update itself is typically the
+  result of applying any number of `chainable` transformations.
+
+  Args:
+    params: a tree of parameters.
+    updates: a tree of updates, the tree structure and the shape of the leaf
+    nodes must match that of `params`.
+
+  Returns:
+    Updated parameters, with same structure and shape as `params`.
+  """
+  return tree_multimap(lambda p, u: p + u, params, updates)
+
+
+### Aliases for popular optimizers. ###
```

Comment:
Are there one or two (maybe less popular) optimizers that you can add as examples of the benefits of composability? For instance, perhaps `nadam` from https://openreview.net/pdf?id=OM0jvwB8jIp57ZJjtNEZ or LARS can be expressed using `chainer` and the existing primitives?

Reply:
As an example I added a `noisy_sgd` as from the paper: https://arxiv.org/abs/1511.06807.

Crucially though, the composable nature of `optix` would allow the user to build in a one liner also a `noisy_adam` or `noisy_rmsprop`, or more generally combine the idea from the paper with any optimizer of his/her choice.

As a comparison to `TF` or `jax/experimental/optimizers.py`, there the user could only add noise to the gradient before applying the adam/rmsprop rescaling, because the adam/rmsprop would immediately apply the updates and the user could not insert itself before the update without rewriting the entire optimizer. 

Instead in optix the user may experiment with adding the noise before or after the adam/rmsprop rescaling (and I suspect the latter, very complicated without optix, will actually be better).